### DISCUSSION TITLE
Mocking and stubbing dependencies in server-side code during e2e tests

### DISCUSSION THREAD

**Discussion by dklymenk** [vercel/next.js#33356] [Help]
While it is not officially documented on the [testing doc](https://nextjs.org/docs/testing), the [with-msw](https://github.com/vercel/next.js/tree/canary/examples/with-msw) example features how to mock http requests. This covers the case when you are simply using next.js to fetch and render data from a CMS like WordPress, but I failed to find any guides or discussions covering the topics of mocking generic dependencies in e2e tests.

Here are some of the potential use cases where you might want to mock something other than an http request in an e2e test:

1. Your next.js app interacts with a DBMS directly.
It is not an uncommon use case. It is also one mentioned in the documentation. While it can be argued that e2e tests need to cover everything from user click to persistence layer, mocking and stubbing database abstraction layer is a common practice since you don't really care about testing code that isn't yours, especially if it's not something between user and your app.
2. Your next.js app interacts with file system.
While not a common use case, [it is featured in official next.js learn course](https://nextjs.org/learn/basics/dynamic-routes/render-markdown). In a typical node application I would stub the relevant methods from `fs` and/or the library that does uses `fs` under the hood and call it a day.
3. Your next.js app sends http requests using some third party library.
This can be anything from an aws s3 sdk to stripe. These libraries send normal http requests and you can mock them with msw, but that would mean that you're essentially testing those libraries and not your own code. Additionally, if you update that library to a new major version and the javascript api changes, but under the hood it is still sending the same request, your test will pass, but in production you would get a runtime error. In all likelihood, if you're using typescript, it would fail to compile, but if we turn it the other way around (the api changes, but the library itself works), the tests would fail, but the production would code work flawlessly. So you need to mock the actual library.
4. Your next.js app sends emails or does any other non-http communication.
Assuming your test sends an actual email over smtp, the only logical way to verify that it was delivered would be to wait one minute, login to the recipients mailbox with a nodejs imap client and verify that the message was delivered. I don't think anyone actually does that. You need to mock your smtp library.

Possible workarounds:
* Use actual test database, maybe replace it with an in-memory one to make things faster.
This would require you to provide credentials for the test database in `.env.test` and to make sure your test runner passes `NODE_ENV=test` to build and start commands. You will also need to care about resetting the database after every test, if your tests update any records or documents. This workaround allows you to adjust the database state before each test (to test the cases when there is data and there is no data), if you are willing to initialise your database client during tests.
* Move all the non-http logic to a separate micro-service and mock your communication with it via msw.
This is probably the best solution for me so far as it allows you to have the highest confidence in your software system (assuming you are testing that micro-service too). This comes at a cost of development time and, if you are unable to make the client-side code to send requests directly to that micro-service, a delay in user interactions introduced by the next.js app relaying the requests.
* Add dummy data directly to the production code.
Assuming you have a function `getProducts` that fetches data from a database or parses json or a csv file, you can add an if statement that would check `APP_ENV === 'test'` and return some fake data.
This doesn't allow you customise the returned value on per-test basis, but does allow you tweak it depending on parameters that are passed to the function.

Just for reference, these are the usual tools for mocking dependencies, would be great if there was a way to use these or a similar library with next.js:
* [Sinon.JS](https://sinonjs.org/releases/v12.0.1/) - supports mocks, spies, stubs and fakes.
* [proxyquire](https://www.npmjs.com/package/proxyquire) - not as versatile as SinonJS, but gets the job done.

Thanks.

**Comment by leahtopchik** [comment]
Agreed - it would be great if Next would support Sinon for mocking/stubbing 3rd party libs

**Comment by helderberto** [comment]
My two cents on this: If you need to mock methods from an external library (I'm using [LaunchDarkly](https://docs.launchdarkly.com/sdk/client-side/javascript#initialize-the-client) in that case), you can create a file that exports these methods and stub these methods in your project, e.g.:

```js
import * as LaunchDarklySDK from '../launch-darkly-lib';

module('service:launch-darkly', function(hooks) {
  setupRenderingTest(hooks);

  hooks.beforeEach(() => {
    this.sandbox = sinon.createSandbox();

    // Stub the initialize method to return a mocked client object
    this.initializeStub = this.sandbox.stub(LaunchDarklySDK, 'initialize').returns({
      on: this.sandbox.stub().resolves(),
      waitUntilReady: this.sandbox.stub(),
    });
  });

  hooks.afterEach(() => {
    // Restore the sandbox to clean up any stubs or mocks
    this.sandbox.restore();
  });

  ...
});
```
